DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/niosv/software/simple_dma_test/src/simple_dma_test.c

```diff
@@ -0,0 +1,219 @@
+  // padding upper 32 bits to all zeros
+  IOWR_32DIRECT(REG_ARG_SOURCE_BASE + 4, 0, 0);
+
+  // DMA destination
+  IOWR_32DIRECT(REG_ARG_DEST_BASE, 0, (unsigned int)destination);
+  // padding upper 32 bits to all zeros
+  IOWR_32DIRECT(REG_ARG_DEST_BASE + 4, 0, 0);
+
+  // DMA length
+  IOWR_32DIRECT(REG_ARG_LENGTH_BASE, 0, BUFFER_LENGTH);
+
+  // DMA start
+  IOWR_32DIRECT(REG_START_BASE, 0, 1);
+
+  // The DMA kernel should immediately start at this point
+}
+
+/// @brief Exercise the DMA kernel.
+///
+/// The test has the following phases: \n
+///
+/// 1. populate a source buffer with an incrementing pattern of 'BUFFER_LENGTH'
+/// bytes \n
+///
+/// 2. clear out the destination buffer \n
+///
+/// 3. instruct the DMA kernel to perform the source --> destination transfer \n
+///
+/// 4. check that the destination contents match the source
+///
+/// @return 0 if the test passes.
+int test_simple_dma() {
+  // allocating source and destination buffers at compile time so that if too
+  // large a value is set for BUFFER_LENGTH then we'll find out early at compile
+  // time
+  unsigned int source[BUFFER_LENGTH / 4];
+  unsigned int destination[BUFFER_LENGTH / 4];
+  int i;
+
+  // initialize the source buffer with an incrementing pattern and clear out the
+  // destination before the accelerator clobbers it
+  for (i = 0; i < (BUFFER_LENGTH / 4); i++) {
+    source[i] = i;
+    destination[i] = 0;
+  }
+
+  // main memory (code_data_ram) is *not* in a peripheral region, so all the
+  // writes to the source and destination need to be flushed from the data cache
+  // to avoid cache coherency issues when the accelerator attempts to access
+  // memory.
+
+  // make sure all that source data that was set gets flushed out to main memory
+  alt_dcache_flush(source, BUFFER_LENGTH);
+
+  // make sure all that destination data that was zeroed out gets flushed out to
+  // main memory
+  alt_dcache_flush(destination, BUFFER_LENGTH);
+
+  //  Configure and start the DMA kernel
+  configure_and_start_dma(source, destination, BUFFER_LENGTH);
+
+  // spin waiting for the accelerator to complete (kernel will fire off
+  // interrupt as well but there is no register as of 2023.1.0 to clear it)
+  while ((IORD_32DIRECT(REG_STATUS, 0) & KERNEL_REGISTER_MAP_DONE_MASK) !=
+         KERNEL_REGISTER_MAP_DONE_MASK) {
+  }
+
+  // Now that the accelerator is done, test the destination buffer for
+  // correctness. Since the destination buffer was already flushed from the data
+  // cache, software can safely read the destination without additional data
+  // flushes. Since the source was previously flushed as well, it will get
+  // fetched from main memory and warm up the data cache just like the
+  // destination buffer in the correctness loop.
+
+  // test the results at the destination buffer, if a failure is detected set
+  // pass to 0 and stop testing
+  for (i = 0; i < (BUFFER_LENGTH / 4); i++) {
```

Comment:
If the kernel code is change so that the BUFFER_LENGTH/4 is outside the loop then I recommend making the same change here.